File: T-Bank/Algorithms-and-Data-Structures-2025/6sprint-Graph1/contest/E-Теория_чисел/solution.py
import sys
import heapq


def solution() -> None:
    """
    Time Complexity O(k)
    Memory Complexity O(k)
    """
    k: int = int(sys.stdin.readline().rstrip())
    min_sum: list[int] = [float('inf')] * k
    min_heap: list[tuple[int, int]] = []
    for num in range(1, 10):
        remainder: int = num % k
        if min_sum[remainder] > num:
            min_sum[remainder] = num
            heapq.heappush(min_heap, (num, remainder))
    while min_heap:
        current_sum, current_remainder = heapq.heappop(min_heap)
        if current_remainder == 0:
            sys.stdout.write('{}\n'.format(current_sum))
            return
        for num in range(10):
            new_sum: int = current_sum + num
            remainder: int = (current_remainder * 10 + num) % k 
            if min_sum[remainder] > new_sum:
                min_sum[remainder] = new_sum
                heapq.heappush(min_heap, (new_sum, remainder))


if __name__ == '__main__':
    solution()

File: T-Bank/Algorithms-and-Data-Structures-2025/6sprint-Graph1/contest/E-Теория_чисел/solution.cpp
#include <iostream>
#include <queue>
#include <vector>
#include <climits>

using namespace std;

void solution() {
    /*
    Time Complexity O(k)
    Memory Complexity O(k)
    */
    long long k;
    scanf("%lld", &k);
    vector<long long> minSum(k, LLONG_MAX);
    priority_queue<pair<long long, long long>, vector<pair<long long, long long>>, greater<pair<long long, long long>>> pq;
    for (long long num = 1; num < 10; ++num) {
        long long remainder = num % k;
        if (num < minSum[remainder]) {
            minSum[remainder] = num;
            pq.push({num, remainder});
        }
    }

    while (!pq.empty()) {
        long long currentSum = pq.top().first;
        long long currentRemainder = pq.top().second;
        pq.pop();
        if (currentRemainder == 0) {
            printf("%lld\n", currentSum);
            return;
        }
        for (long long digit = 0; digit < 10; ++digit) {
            long long newSum = currentSum + digit;
            long long newRemainder = (currentRemainder * 10 + digit) % k;
            if (newSum < minSum[newRemainder]) {
                minSum[newRemainder] = newSum;
                pq.push({newSum, newRemainder});
            }
        }
    }
}

int main() {
    solution();
}